pmd-ui/src/main/java/net/sourceforge/pmd/util/fxdesigner/XPathPanelController.java

```diff
@@ -113,9 +120,72 @@ public void initialize(URL location, ResourceBundle resources) {
                            // Reevaluate XPath anytime the expression or the XPath version changes
                            .or(xpathVersionProperty().changes())
                            .subscribe(tick -> parent.refreshXPathResults());
+
+        xpathExpressionArea.plainTextChanges()
+                           .filter(t -> (t.getInserted().matches("[a-zA-Z/]")))
+                           .subscribe(t -> {
+                               try {
+                                   if (xpathExpressionArea.getText().contains("/")) {
+                                       if (!t.getInserted().equals("/")) {
+                                           autoComplete(xpathExpressionArea.getText().substring(xpathExpressionArea.getText().lastIndexOf("/")));
+                                       }
+                                   }
+                               } catch (IOException e) {
+                                   e.printStackTrace();
+                               } catch (ClassNotFoundException e) {
+                                   e.printStackTrace();
+                               }
+                           });
+    }
+
+    private void autoComplete(String s) throws IOException, ClassNotFoundException {
+
+        autoCompletePopup = new ContextMenu();
+        List<MenuItem> resultToDisplay = new ArrayList<>();
+
+        String language = parent.getLanguageVersion().getName().replaceAll("[0-9]", "").replaceAll("//s", "").toLowerCase().trim();
+        XPathSuggestions xPathSuggestions = new XPathSuggestions(language);
+
+        List<String> suggestions = xPathSuggestions.getXPathSuggestions();
```

Comment:
The XPathSuggestions should encapsulate the algorithm to find suggestions from the input string. The call should look like `xpathSuggestions.getSuggestions(input)` and return only the suggestions that match the string, the definiton of "x matches s" being defined by that procedure. 

Then you could for example stream the list, map it to menu items, and take only the first 5 to not overwhelm the user.